### PR TITLE
Replace IO by polymorphic F

### PR DESCRIPTION
**Pull request by Fristi**
Such that we can use it with ZIO or Monix

### CODE REVIEW

**Comment by regis-leray** [comment]
Thanks for your contribution. But sadly this library cannot be built to support any `Stream` implementation, sadly there is no common API interface. That is why i never use polymorph F

There is already a project to support `ZIO-FTP` contributed by me :) which expose ZStream :)

**Comment by Fristi** [comment]
Why would that not work? A `zio.Task` implements `ConcurrentEffect`, the tests pass :) What is the problem?

**Comment by Fristi** [comment]
I like your library zio-ftp, but I don't want to use it. I use fs2 everywhere for streams (http4s, doobie and even paginated api calls). My core module depends on it and on my core module all other modules depend. Introducing zio in the core for me is a no-go, as I already locked myself in to fs2. There might be benefits to using zio-streams, but in this case it's not desirable

**Comment by regis-leray** [comment]
There is no common API between fs2.Stream and ZStream 
👇 
```scala
fs2.Stream[F, Byte]
```

Can you explain the problem you are facing with the current implementation ?

**Comment by Fristi** [comment]
You can use a fs2 Stream with a `zio.Task` like `fs2.Stream[Task, Byte]`. 

Indeed there is no conversion between fs2 Stream and ZStream. 

I think it's oke to have a fs2-ftp and zio-ftp. You can either go for fs2-ftp if you want to pick your effect type. The standard cats.effect.IO. Monix or zio.Task. These will all work with fs2.Stream. For opinionated stacks which favor ZIO for everything I think zio-ftp is very easy to setup. 

I don't want it, because my core module has a low dependency footprint. It depends on refined, cats and fs2. It only contains tagless final interfaces and models which are implemented in other modules in terms of doobie, sttp and such. In higher layers (service layer and the api layer I use ZIO as it's the winner in effect monads at the moment).

**Comment by Fristi** [comment]
ZIO has `zio-interop-cats` which exports `Sync`, `ConcurrentEffect` instances and other cats-effect instances for `ZIO[R, E, A]`

**Comment by regis-leray** [comment]
Let me clarify my thoughts on polymorph F  in fs2-ftp : 

- `F[_]` is the ability to represent an effect which is tight by some api interface (trait). I cannot make totally generic this library since there is no common API between fs2.Stream and zio.Stream and choose which stream implementation i want at a method declaration level (which was my intention in the first place, not to support 2 libraries). 

- `F[_]` bring more complexity than it should be, an also less headache in my opinion, not fan of tagless final.

- I made the effort to build zio-ftp now you can choose what implementation you might prefer.

if you still want to mix fs2.Stream[ZIO]  you can lift zio into cats by calling explicitly `zio-interop-cats` at you state

So i prefer to stick to `cats.effect.IO` for this reason. thanks again for your contribution.
And you can always fork it :)

**Comment by Fristi** [comment]
> I made the effort to build zio-ftp you can choose what implementation/style you might prefer.

I think a lot of users are thankful for that :) 

> F[_] bring more complexity than it should be, an also more headache in my opinion, not fan of tagless final.

It's a bit more complex, yet I think if you look at the changes I've made it's quite minimal. I agree a concrete effect type is easier to understand yet I think there is also a _trend_ of getting rid of this style while this might be a good reason to pursue it in this case.





> F[_] is the ability to represent an effect which is tight by some api interface (trait). I cannot make totally generic this library since there is no common API between fs2.Stream and zio.Stream.

There is indeed no common interface of fs2 and ZStream. fs2 was _not_ build on `cats.effect.IO`, but rather on the type classes cats-effect offers such as `Sync` and `ConcurrentEffect` and so on. Therefore if you look at libraries which use fs2 to talk to datasources they are 90% implemented in terms of a polymorphic `F[_]` with type class constraints on it. This gives the flexibility to plugin `zio.Task`, a monix effect type and the standard `cats.effect.IO` type. 

> So i prefer to stick to cats.effect.IO for this reason. thanks again for your contribution.
And you can always fork it :)

What's your plan with this library? Would you deprecate it in favor of ZIO? I don't like to have forks, I would rather work together. I think the work you did here is great, would be a shame to have forks or leave it to die.

**Comment by regis-leray** [comment]
After some thoughts, since some people might be facing same problem willing to support any effects i will merge your PR can you please update the docs.

File: src/main/scala/ray/fs2/ftp/Ftp.scala [hunk 1]
```diff
@@ -2,45 +2,50 @@ package ray.fs2.ftp
...
-import cats.syntax.monadError._
+import cats.effect.{ Blocker, ConcurrentEffect, ContextShift, Resource }
+import cats.implicits._
 import fs2.Stream
 import org.apache.commons.net.ftp.{ FTP, FTPSClient, FTPClient => JFTPClient }
 import ray.fs2.ftp.FtpSettings.UnsecureFtpSettings
 
-final private class Ftp(unsafeClient: JFTPClient, blocker: Blocker) extends FtpClient[JFTPClient] {
+final private class Ftp[F[_]](unsafeClient: JFTPClient, blocker: Blocker)(
+  implicit CE: ConcurrentEffect[F],
```
[the remark this answers is not in the document]
Comment (Fristi) [reply src/main/scala/ray/fs2/ftp/Ftp.scala:12]: I see your point. In most cases you want to know on which Executors this stuff runs and therefore in the examples it either uses the `ContextShift` from `IOApp` or a custom defined one.

Also I like to omit this constraint from the `Ftp` interfaces , such that it easy to mock now

**Comment by Fristi** [comment]
Thanks! I updated the docs, if there is anything else let me know

**Comment by regis-leray** [comment]
thanks a lot

**Pull request merged**